### ISSUE TITLE
List Not Update

### ISSUE THREAD

**Issue by tomeroto** [mikepenz/FastAdapter#338]
Hi everyone, I was follow the all API step by step using fastadapter. I have some problem.
When I search some query and update the item, then on the list is not update (until I touch the screen).

I use search view:

                @Override
                public boolean onQueryTextSubmit(String query) {

                    if(!querySearch.equals(query)){
                        querySearch = query; // the query search

                        // 1. First, clear the array of data
                        fastItemAdapter.clear();

                        // 2. Notify the adapter of the update
                        fastItemAdapter.notifyDataSetChanged();
                        // fastItemAdapter.notifyAdapterDataSetChanged();

                        // 3. Reset endless scroll listener when performing a new load
                        endlessRecyclerOnScrollListener.resetPageCount();
                    }

                    return true;
                }

The Layout Grid:

       GridLayoutManager gridLayoutManager = new GridLayoutManager(ChooseUserActivity.this, 3);
        gridLayoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override
            public int getSpanSize(int position) {
                switch (fastItemAdapter.getItemViewType(position)) {
                    case R.id.fastadapter_progress_item_id:
                        return 3;
                    case R.id.fastadapter_user_item_id:
                        return 1;
                    default:
                        return -1;
                }
            }
        });

Wrapper:

       recyclerView.setAdapter(footerAdapter.wrap(fastItemAdapter));

I Dont how to fix it, because I was log the request URL and it response the data, but on the list (recyleview) somethimes not showing (until I touch or scroll the screen)

*NOTE: when I remove the footerAdapter (not using footer adapter) then it working good. But I need loading animation on footer so I use footer Adapter :'(

**Comment by mikepenz** [comment]
@tomeroto are you updating the adapter on the UI thread?

**Comment by tomeroto** [comment]
this is my full Activity:

```
public class ChooseUserActivity extends AppCompatActivity {

    private FastItemAdapter<UserItem> fastItemAdapter;
    private FooterAdapter<ProgressItem> footerAdapter;
    private RecyclerView recyclerView;
    private EndlessRecyclerOnScrollListener endlessRecyclerOnScrollListener;
    private SwipeRefreshLayout swipeRefreshLayout;

    private String querySearch = "";
    private String memberType;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_choose_user);

        String title = getIntent().getStringExtra("title");
        memberType = getIntent().getStringExtra("member_type");

        // Find the toolbar view inside the activity layout
        TitleMarqueeToolbar toolbar = (TitleMarqueeToolbar) findViewById(R.id.toolbar);
        // Sets the Toolbar to act as the ActionBar for this Activity window.
        // Make sure the toolbar exists in the activity and is not null
        try {
            setSupportActionBar(toolbar);
        } catch (Throwable t) {
            // WTF SAMSUNG!!!
        }

        // Display icon in the toolbar
        // getSupportActionBar().setDisplayShowHomeEnabled(true);
        // getSupportActionBar().setLogo(R.mipmap.ic_launcher);
        // getSupportActionBar().setDisplayUseLogoEnabled(true);

        ActionBar actionBar = getSupportActionBar();
        // Remove default title text (app name)
        actionBar.setDisplayShowTitleEnabled(false);
        // show back button
        actionBar.setDisplayHomeAsUpEnabled(true);
        // make marquee title bar
        toolbar.setTitle(title);

        // swipe
        swipeRefreshLayout = (SwipeRefreshLayout) findViewById(R.id.swipeRefreshLayout);

        //create our FastAdapter which will manage everything
        fastItemAdapter = new FastItemAdapter<>();
        fastItemAdapter.withSelectable(true);
        fastItemAdapter.withItemEvent(new UserItem.UserClickEvent());

        //create our FooterAdapter which will manage the progress items
        footerAdapter = new FooterAdapter<>();

        //configure our fastAdapter
        fastItemAdapter.withOnClickListener(new FastAdapter.OnClickListener<UserItem>() {
            @Override
            public boolean onClick(View v, IAdapter<UserItem> adapter, UserItem item, int position) {

                Intent returnIntent = getIntent();
                returnIntent.putExtra("id", item.getId());
                returnIntent.putExtra("name", item.getFirstName() + (item.getLastName() == null ? "" : " " + item.getLastName()));
                setResult(Activity.RESULT_OK, returnIntent);
                finish();// finishing activity

                return false;
            }
        });

        GridLayoutManager gridLayoutManager = new GridLayoutManager(ChooseUserActivity.this, 3);
        gridLayoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override
            public int getSpanSize(int position) {
                switch (fastItemAdapter.getItemViewType(position)) {
                    case R.id.fastadapter_progress_item_id:
                        return 3;
                    case R.id.fastadapter_user_item_id:
                        return 1;
                    default:
                        return -1;
                }
            }
        });
        recyclerView = (RecyclerView) findViewById(R.id.recyclerview);
        recyclerView.setLayoutManager(gridLayoutManager);
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        //set our adapters to the RecyclerView
        //we wrap our FastAdapter inside the ItemAdapter -> This allows us to chain adapters for more complex useCases
        recyclerView.setAdapter(footerAdapter.wrap(fastItemAdapter));

        endlessRecyclerOnScrollListener = new EndlessRecyclerOnScrollListener(footerAdapter, Api.MAX_RECORD_PER_PAGE) {
            @Override
            public void onLoadMore(final int currentPage) {
                loadData(querySearch, currentPage);

                //simulate networking (2 seconds)
                /*Handler handler = new Handler();
                handler.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        loadData(currentPage);
                    }
                }, 2000);*/
            }
        };
        recyclerView.addOnScrollListener(endlessRecyclerOnScrollListener);

        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                // direct hide, because we have footer loading
                swipeRefreshLayout.setRefreshing(false);

                // 1. First, clear the array of data
                fastItemAdapter.clear();

                // 2. Notify the adapter of the update
                fastItemAdapter.notifyDataSetChanged();

                // 3. Reset endless scroll listener when performing a new load
                endlessRecyclerOnScrollListener.resetPageCount();
            }
        });

        /**
         * Showing Swipe Refresh animation on activity create
         * As animation won't start on onCreate, post runnable is used
         */
        swipeRefreshLayout.post(new Runnable() {
            @Override
            public void run() {
                // show loading
                // swipeRefreshLayout.setRefreshing(true); // disable loading for first time

                //load first items to ItemAdapter
                endlessRecyclerOnScrollListener.resetPageCount();
            }
        });

        //restore selections (this has to be done after the items were added
        fastItemAdapter.withSavedInstanceState(savedInstanceState);
    }

    @Override
    protected void onStop() {
        Ion.getDefault(ChooseUserActivity.this).cancelAll(ChooseUserActivity.this);

        super.onStop();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        //add the values which need to be saved from the adapter to the bundel
        outState = fastItemAdapter.saveInstanceState(outState);
        super.onSaveInstanceState(outState);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        int id = item.getItemId();

        switch (id) {
            case R.id.action_search:
                // Not implemented here
                return false;
            case android.R.id.home:
                this.finish();
                return true;
            default:
                break;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu items for use in the action bar
        getMenuInflater().inflate(R.menu.general_search, menu);

        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
            final SearchView searchView = (SearchView) menu.findItem(R.id.action_search).getActionView();
            searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
                @Override
                public boolean onQueryTextSubmit(String query) {

                    // TODO: I DON"T WHY IS DOUBLE SEARCH
                    if(!querySearch.equals(query)){
                        querySearch = query;

                        // 1. First, clear the array of data
                        fastItemAdapter.clear();

                        // 2. Notify the adapter of the update
                        fastItemAdapter.notifyDataSetChanged();
                        // fastItemAdapter.notifyAdapterDataSetChanged();

                        // 3. Reset endless scroll listener when performing a new load
                        endlessRecyclerOnScrollListener.resetPageCount();
                    }

                    return true;
                }

                @Override
                public boolean onQueryTextChange(String query) {

                    if(query.isEmpty()){
                        if(!querySearch.isEmpty()){
                            querySearch = query;

                            // 1. First, clear the array of data
                            fastItemAdapter.clear();

                            // 2. Notify the adapter of the update
                            fastItemAdapter.notifyDataSetChanged();
                            // fastItemAdapter.notifyAdapterDataSetChanged();

                            // 3. Reset endless scroll listener when performing a new load
                            endlessRecyclerOnScrollListener.resetPageCount();
                        }
                    }

                    return true;
                }
            });
        } else {
            menu.findItem(R.id.action_search).setVisible(false);
        }


        return super.onCreateOptionsMenu(menu);
    }

    private void loadData(String query, Integer page) {

        // load if online
        // if (Common.isOnline(getContext())) {

        // show loading footer
        footerAdapter.clear(); // make sure
        footerAdapter.add(new ProgressItem().withEnabled(false));

        // make URI GET PARAMS
        String uri = Uri.parse(Api.GET_USERS)
                .buildUpon()
                .appendQueryParameter("q", query)
                .appendQueryParameter("group_name", getString(R.string.app_name))
                .appendQueryParameter("member_type", memberType) // TODO: CHANGE TO VARS DON'T HARDCODE
                .appendQueryParameter("p", page.toString())
                .build().toString();

        // get information from URL
        Api.doGET(ChooseUserActivity.this, uri, new FutureCallback<Response<String>>() {
            @Override
            public void onCompleted(Exception ex, Response<String> response) {

                // hide loading footer
                footerAdapter.clear();

                // has error after process
                if (ex != null) {
                    // break
                    return;
                }

                // If valid response
                if (response.getHeaders().code() == 200) {

                    try {
                        JSONObject responseJsonObject = new JSONObject(response.getResult());
                        Integer statusCode = 0;
                        String message = "";
                        if (responseJsonObject.has("status_code")) {
                            statusCode = responseJsonObject.getInt("status_code");
                        }
                        if (responseJsonObject.has("message")) {
                            message = responseJsonObject.getString("message");
                        }

                        if (statusCode == 200) {
                            if (responseJsonObject.has("data")) {
                                JSONArray jsonArray = responseJsonObject.getJSONArray("data");

                                for (int i = 0; i < jsonArray.length(); i++) {

                                    JSONObject jsonObject = jsonArray.getJSONObject(i);

                                    fastItemAdapter.add(fastItemAdapter.getAdapterItemCount(), new UserItem()
                                            .withIdentifier(jsonObject.getInt("id"))
                                            .setId(jsonObject.getInt("id"))
                                            .setUsername(Common.JsonObjectStringNullReplacer(jsonObject, "username"))
                                            .setFirstName(Common.JsonObjectStringNullReplacer(jsonObject, "first_name"))
                                            .setLastName(Common.JsonObjectStringNullReplacer(jsonObject, "last_name"))
                                            .setGender(Common.JsonObjectStringNullReplacer(jsonObject, "gender"))
                                            .setBirthDate(Common.JsonObjectStringNullReplacer(jsonObject, "birth_date"))
                                            .setPlaceOfBirth(Common.JsonObjectStringNullReplacer(jsonObject, "place_of_birth"))
                                            .setReligion(Common.JsonObjectStringNullReplacer(jsonObject, "religion"))
                                            .setEmail(Common.JsonObjectStringNullReplacer(jsonObject, "email"))
                                            .setPhone(Common.JsonObjectStringNullReplacer(jsonObject, "phone"))
                                            .setAdditionalPhone(Common.JsonObjectStringNullReplacer(jsonObject, "additional_phone"))
                                            .setAddress(Common.JsonObjectStringNullReplacer(jsonObject, "address"))
                                            .setCountryId(jsonObject.getInt("country_id"))
                                            .setCountry(Common.JsonObjectStringNullReplacer(jsonObject, "country"))
                                            .setProvince(Common.JsonObjectStringNullReplacer(jsonObject, "province"))
                                            .setRegency(Common.JsonObjectStringNullReplacer(jsonObject, "regency"))
                                            .setVillage(Common.JsonObjectStringNullReplacer(jsonObject, "village"))
                                            .setPhotoFilename(Common.JsonObjectStringNullReplacer(jsonObject, "photo_filename"))
                                            .setDistrict(Common.JsonObjectStringNullReplacer(jsonObject, "district"))
                                            .setSumRating(jsonObject.getDouble("sum_rating"))
                                            .setCountUserRating(jsonObject.getInt("count_user_rating"))
                                    );
                                }
                            }
                        } else {
                            ColoredSnackbar.alert(Snackbar.make(ChooseUserActivity.this.findViewById(android.R.id.content), message, Snackbar.LENGTH_LONG)).show();
                        }
                    } catch (JSONException e) {
                        ColoredSnackbar.warning(Snackbar.make(ChooseUserActivity.this.findViewById(android.R.id.content), getString(R.string.connection_error), Snackbar.LENGTH_LONG)).show();
                    }
                }
            }
        });
    }
}
```
I was follow an example from your step by step but still not working

**Comment by mikepenz** [comment]
@tomeroto possibly this happens because you do the update inside submit and textchange?

**Comment by tomeroto** [comment]
yes, on my activity, it have search query and do update the items after submiting. This part of code from above:

             ```
public boolean onQueryTextSubmit(String query) {

                    // TODO: I DON"T WHY IS DOUBLE SEARCH
                    if(!querySearch.equals(query)){
                        querySearch = query;

                        // 1. First, clear the array of items
                        fastItemAdapter.clear();

                        // 2. Notify the adapter of the update (I dont know what the best of call for this)
                        fastItemAdapter.notifyDataSetChanged();
                        // fastItemAdapter.notifyAdapterDataSetChanged();

                        // 3. By call this method, then we load new items with or without quer
                        // it will call method :  private void loadData(String query, Integer page)
                        endlessRecyclerOnScrollListener.resetPageCount();
                    }

                    return true;
                }

```
But after I update the items on fastitemadapter, why it not show update on recycleview. And why after I remove footerAdapter (not using footerAdapter) it run good. I think problem is between endlessscroll and headeradapter/footeradapter. But I don't know how to fix it. I just want to use footeradapter to show loading animation until the data is loaded. :'(

**Comment by mikepenz** [comment]
could you make sure the `loadData()` method is executed on the main thread?

**Comment by tomeroto** [comment]
Yes I was call the loadData() method on main thread and still not working until I remove the footerAdapter.
If the method of loadData have problem, but why is working good when I remove the footerAdapter.

So I change adapter from this:

`recyclerView.setAdapter(footerAdapter.wrap(fastItemAdapter));`

to this:

`recyclerView.setAdapter(fastItemAdapter);`

Do you have an example using fastadapter with footeradapter and search/query data from internet?

**Comment by mikepenz** [comment]
@tomeroto that'S weird. as the footeradapter works in all projects including the sample I have. 

as for your sample. There is none, as nothing special on adapter side is required. Just add the items to the adapter via the provided methods. done.

**Comment by tomeroto** [comment]
confusing...

I don't know how to fix it.

Sometimes, after I clear and add several items on adapter and call:
`fastItemAdapter.notifyDataSetChanged();`

but list on recycleview is not update.  :(

**Comment by mikepenz** [comment]
@tomeroto the items and the notify methods are already called internally via the `FastAdapteR` so you should not have to do this anymore. 

It's weird as it works for all cases where I use the `FastAdapter` and I even have multiple sub adapters

**Comment by mikepenz** [comment]
Do you have an update for this? 

By the way `bindView` is handled and called by the `Adapter` via the `RecyclerView`. So if the `RecyclerView` does not need to re-bind it won't call this method

**Comment by tomeroto** [comment]
I no have idea.
And I was search all your examples, and I don't see an example using query search with online update data.
On your example, you just load all data to adapter and search/filter only from available/loaded data on adapter.

I just want to do:

1. Input query, and submit to server via web service
2. When data return from server, I clear old data on adapter, and insert new data to adapter
3. And as you say fastadapter will auto show the data on the list (recycleview)

but sometimes when server return the data, it not update to list (recycleview).

I will try to solve this as soon as possible.

**Comment by mikepenz** [comment]
@tomeroto you could send me the project (remove everything which is cirtical for me to see) via e-mail. And I can have a look on that exact part. 

The only thing I can say is, if you set the new items to the adapter on the main thread, it will work. If it does not it is not related to the `FastAdapter` but to some other issue with the RecyclerView, as the adapter only holds the items and notifies the adapter, it is not the component of the RV handling animations, or updating the list. That is all done via the LayoutManater + the Animator which are coordinated from the RV

**Comment by tomeroto** [comment]
Hi Mike, I just send you an example of source to show you the problem that I found. Any way thank you before

**Comment by mikepenz** [comment]
You have my e-mail? If the zip is too large upload it to google drive or so ans share a link please

**Comment by tomeroto** [comment]
Yes I send to : mikepenz@gmail.com

**Comment by mikepenz** [comment]
ok I'll have a look

**Comment by tomeroto** [comment]
Not working :(, I was test with two solutions above.
This is link of video to show you the problem:
https://drive.google.com/open?id=0B69BM-MOwRDBMVZOVnZ2b3pOcnM

I hope you understand.

**Comment by mikepenz** [comment]
Ok my initial thought was wrong. it was not related to the ID sorry for that. The ID just masked the issue a bit and improved the behaviour. 

The real issue is that the page size you have set is 10 items. which results in the onLoadMore trigger to fire multiple times very fast. if this combined with the clearing of the list happens (because of fast typing) the RecyclerView will be lost in an unresolveable state. 

The easiest to fix it, ist just to set a higher page size like 24 items or so:

https://drive.google.com/open?id=0B4Y8HN2BnSjpNjFNVjQyZGg3MkE

**Comment by tomeroto** [comment]
oh okay, good info for me, thank you for your support.

**Comment by AllanWang** [comment]
I have a similar issue with this. I load my data using retrofit and I keep my recyclerview inside a swiperefreshlayout. Originally, I only used one fastitemadapter (mAdapter) and everything worked well, animations and everything. Now I wanted to wrap it inside a headeradapter (because I want one header view), but it breaks the notifications. If I refresh the adapter, calling mAdapter.clear() will not show the changes since headerAdapter doesn't know its contents have been modified. I'd also need to use mAdapter.add(), which doesn't produce a result unless I tap somewhere.

[Attached](https://github.com/CTFMcGill/CTF-Android/blob/master/CTF/src/main/java/ca/mcgill/science/ctf/fragments/AccountJobFragment.java) is the fragment containing the adapters. A lot of the contents are in subclasses, but essentially, `BasePrintJobFragment` has an mAdapter that is cleared and reloaded on refresh. I added the headerAdapter as shown.
